text/0000-export-to-oci.md

```diff
@@ -0,0 +1,339 @@
+- Push the application image (exported in OCI layout format) into the Daemon, because that is what users are expecting.
+
+During the proof of concept implementation I choose to use [skopeo](https://github.com/containers/skopeo) tool to solve the problem of interacting with the Daemon. The reason to do it was **simplicity** for the PoC developed but we believe this is a good subject to talk about with the community.
+
+The following workflow was developed:
+- Pack download the [skopeo image](https://github.com/containers/skopeo/blob/main/install.md#container-images), similar as it is downloading the other dependencies (Lifecycle, Buildpacks)
+- Pack executes [skopeo copy](https://github.com/containers/skopeo/blob/main/docs/skopeo-copy.1.md)  command in a container
+  - Copy image from the Daemon into the filesystem, in OCI layout format, before running Lifecycle
+  - Copy image from filesystem into the Daemon after the export phase was executed
+
+The following Dynamic Diagram from the C4 model, can give a little idea of the pieces implemented during the Poc
+
+![](https://i.imgur.com/SkY3l62.png)
+
+### Measuring of performance impact
+
+In order to have an idea on how much is affected the performance of exporting to the Daemon using the OCI layout format, the following metrics were taken.
+
+Using a local workstation with the following specifications:
+- **(MacOS 12.3.1 / 2,4 GHz 8-Core Intel Core i9 / 32 GB 2667 MHz DDR4)**
+
+We built 5 times the Java, Kotlin and Ruby samples codes from our [repository](https://github.com/buildpacks/samples/tree/main/apps) and took the building's average time using the Daemon and the OCI layout format approaches.
+
+The table above summarized the results we got.
+
+![](https://i.imgur.com/zuPZ6Xk.png)
+
+Times are expressed in **seconds** and the first thing we noticed is for Java and Kotlin the `build time` can be affected by the network and the availability of maven repositories, so I decided to take the `same build time` to compare both approaches.
+
+Here are my thoughts about these results:
+- Java and Kotlin behavior are very similar, exporting only to OCI format increases 5% the time compared to Daemon approach, and from user perspective it represents a 5 seconds increase of time.
+- On the other hand for the Ruby application, exporting to OCI format represents a  20% increase of the time but from user perspective it is only 1.5 seconds which is probably difficult to notice from user perspective.
+- When the time spent for Pack to prepare the environment for the lifecycle execution (downloading the run-image from a registry to OCI format) and loading the OCI image from disk to the Daemon (which is the expected behavior from Users) is added, then:
+  - The Java and Kotlin applications time increases was **13%**, representing **+13 seconds** from user perspective
+  - The Ruby application increases **82%** but from user's side it represents **7 seconds**
```

Comment:
Do we know why this increase exists? Are there any potential solutions to improve the performance?

Reply:
I think with the latest table where I summarized the results we can see the difference came from
- Skopeo -> OCI
- OCI -> Skopeo